qstack/spahm/rho/utils.py

```diff
@@ -70,33 +79,141 @@ def get_xyzlist(xyzlistfile):
+                    labels.extend(x[:,0])
             else:
                 reps.extend(x)
         else:
-           if type(x[0]) == str:
-                reps.extend(x[1])
+           if type(x[0]) == str or type(x[0]) == np.str_:
+                reps.append(x[1])
                 labels.extend(x[0])
            else:
-                reps.extend(x)
+                reps.append(x) 
+        if printlevel > 0:
+            i+=1
+            progress.update(i)
+    if printlevel > 0: progress.finish()
     try:
         reps = np.array(reps, dtype=float)
     except:
+        print(len(reps))
+        shapes = [r.shape[0]  for r in reps]
+        shapes = set(shapes)
+        print(shapes)
         raise RuntimeError("Error while loading representations, check the parameters")
-    reps = np.array(reps, ndmin=1)
     if with_labels:
         return reps, labels
     else:
         return reps
+
+def add_progressbar(legend='', max_value=100):
+    import progressbar
+    import time
+    #import logging
+    #progressbar.streams.wrap_stderr()
+    #logging.basicConfig()
+    widgets=[\
+    ' [', progressbar.Timer(), '] ',\
+    progressbar.Bar(),\
+    ' (', progressbar.ETA(), ')']
+    bar = progressbar.ProgressBar(widgets=widgets, max_value=max_value, redirect_stdout=True).start()
+    return bar
+
+def build_reaction(reacts_file, prods_file, local=False, print_level=0, summ=True, diff=True):
+    reacts = []
+    with open(reacts_file, 'r') as r_in:
+        lines = r_in.readlines()
+        for line in lines:
+            line = line.rstrip('\n')
+            structs = line.split(' ')
+            reacts.append(structs)
+    prods = []
+    with open(prods_file, 'r') as p_in:
+        lines = p_in.readlines()
+        for line in lines:
+            line = line.rstrip('\n')
+            structs = line.split(' ')
+            prods.append(structs)
+    tot = len(reacts)+len(prods)
+    if print_level > 0 : progress = add_progressbar(max_value=tot)
+    i = 0
+    XR = []
+    for rxn in reacts:
+        xr = []
+        for r in rxn:
+            xr.append(load_reps(r, from_list=False, with_labels=False, local=local, summ=True if local else False, single=True))
+        xr = np.squeeze(xr)
+#        print(xr.shape)
+#        exit()
+        if summ and xr.ndim > 1:
+            xr = xr.sum(axis=0)
+        XR.append(xr)
+        i+=1
+        if print_level > 0 : progress.update(i)
+    XP = []
+    for rxn in prods:
+        xp=[]
+        for p in rxn:
+            xp.append(load_reps(p, from_list=False, with_labels=False, local=local, summ=True if local else False, single=True))
+        xp = np.squeeze(xp)
+        if summ and xp.ndim > 1:
+            xp = xp.sum(axis=0)
+        XP.append(xp)
+        i+=1
+        if print_level > 0 : progress.update(i)
+    XR = np.squeeze(XR)
+    XP = np.squeeze(XP)
+    if diff : rxn = XP - XR
+    else: rxn = (XR, XP)
+    if print_level > 0 : progress.finish()
+    return rxn
```

Comment:
We can also try to merge it with the reaction-reps wrapper you added ?
or as an alternative ?

Reply:
I'm starting to think that for the reaction stuff we need a separate module/file with the wrapper and dataloader

Reply:
yeah that makes sense!

It will probably also contain the spahm-b for reactions and related stuff!
**I'll revert these commits and just keep the functions in my trashy branch for now.**